In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import urllib.request
from bs4 import BeautifulSoup
from geopy.geocoders import Nominatim
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
import matplotlib.cm as cm# Matplotlib and associated plotting modules
import matplotlib.colors as colors
from sklearn.cluster import KMeans# import k-means from clustering stage
import folium # map rendering library
import os
!pip install uszipcode
from uszipcode import SearchEngine
from datetime import date, timedelta

     |████████████████████████████████| 378 kB 6.9 MB/s 
     |████████████████████████████████| 77 kB 3.2 MB/s 


# Introduction

In this project we will explore San Francisco Neighborhoods and historical health violations in restaurants in each of them. For it we will access the City of San Francisco DataSF website and dataset available. 

We are looking to answer each of the following questions:
* 	What are the most common violations in this city?
* 	Can neighborhoods be clustered based on violation types to ensure effective deployment of marketing and specialists/packages?
* 	What are the most common food venue types in the neighborhoods?


# Step 1 : Getting Neighborhood information

First step will be to scrape the zipcodes and Neigborhoods from  http://www.healthysf.org/bdi/outcomes/zipmap.htm, this in order to obtain the zip codes we will use to find the latitude and longitude data points needed to plot the neighborhoods.

In [2]:
response = requests.get("http://www.healthysf.org/bdi/outcomes/zipmap.htm")
soup = BeautifulSoup(response.text, "lxml")
table = soup.find_all("table")
df = pd.read_html(str(table))
df = pd.DataFrame(df[4])
df.columns = df.iloc[0]
df = df.iloc[1:-1, :-1]
sf_neigh = df
sf_neigh.head()

0 Zip Code                             Neighborhood
1    94102  Hayes Valley/Tenderloin/North of Market
2    94103                          South of Market
3    94107                             Potrero Hill
4    94108                                Chinatown
5    94109             Polk/Russian Hill (Nob Hill)

Now we will use the zip codes to find the needed latitude and longitude. We will use `uszipcode`, a database in Python that comes in handy for this purpose.

In [3]:
search = SearchEngine(simple_zipcode=True)
latitude=[]
longitude=[]

for index, row in sf_neigh.iterrows():
    zipcode= search.by_zipcode(row["Zip Code"]).to_dict()
    latitude.append(zipcode.get("lat"))
    longitude.append(zipcode.get("lng"))
    
sf_neigh["Latitude"] = latitude
sf_neigh["Longitude"] = longitude

sf_neigh.head()

Start downloading data for simple zipcode database, total size 9MB ...
  1 MB finished ...
  2 MB finished ...
  3 MB finished ...
  4 MB finished ...
  5 MB finished ...
  6 MB finished ...
  7 MB finished ...
  8 MB finished ...
  9 MB finished ...
  10 MB finished ...
  Complete!


0 Zip Code                             Neighborhood  Latitude  Longitude
1    94102  Hayes Valley/Tenderloin/North of Market    37.780   -122.420
2    94103                          South of Market    37.780   -122.410
3    94107                             Potrero Hill    37.770   -122.390
4    94108                                Chinatown    37.791   -122.409
5    94109             Polk/Russian Hill (Nob Hill)    37.790   -122.420

During early testing of mapping I noticed there is an issue with the location returned by `uszipcode` for the Outer Richmond neighborhood. I'll display the value returned by the python library. We will update this location to reflect correctly in the map

In [4]:
sf_neigh.loc[sf_neigh["Zip Code"]=='94121']

0  Zip Code    Neighborhood  Latitude  Longitude
13    94121  Outer Richmond      37.8     -122.7

In [5]:
sf_neigh.loc[sf_neigh["Zip Code"]=='94121']=[['94121','Outer Richmond',37.781934,-122.4906]]
sf_neigh.loc[sf_neigh["Zip Code"]=='94121']

0  Zip Code    Neighborhood   Latitude  Longitude
13    94121  Outer Richmond  37.781934  -122.4906

# Step 2 :  Create a map of San Francisco showing its Neighborhoods

We will use geopy to get the latitude and longitude for San Francisco

In [6]:
#I was using geopy originally but it has proven to be unstable. Hard coding to ensure stability and availability
latitude = 37.7749
longitude = -122.4194
print('The geograpical coordinates of San Francisco are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of San Francisco are 37.7749, -122.4194.


We will then use folium to plot the map and overlay the neighborhood markers

In [7]:
sf_df = sf_neigh
map_sf = folium.Map(location = [latitude, longitude], zoom_start=12, height=500, width=600)

# add markers to map
for lat, lng, neighborhood in zip(sf_df['Latitude'], sf_df['Longitude'], sf_df['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
        [lat, lng],
        radius = 10,
        popup = label,
        color = 'black',
        fill = True,
        fill_color = '#696969',
        fill_opacity = 0.7,
        parse_html = False).add_to(map_sf)  
        
map_sf

# Step 3 : Data Exploration and Preparation for the violations dataset

We will use the Restaurant Scores dataset available at https://datasf.org/. The dataset contains records from October 2016 to November 2019. We will refer to data from 2019 only as it will be most relevant for our purposes.

Pull the dataset into a dataframe:

In [8]:
violations_df = pd.read_csv('https://data.sfgov.org/api/views/pyih-qa8i/rows.csv')
violations_df.shape

(53973, 23)

Drop all records older than January 1st, 2019. Drop columns we do not need. Drop rows with no zipcode or violation. Reset Index

In [9]:
violations_df['inspection_date'] = pd.to_datetime(violations_df['inspection_date'])
cutoff = '01/01/2019'
violations_df = violations_df.loc[(violations_df['inspection_date'] > cutoff)]
violations_df = violations_df.drop(['business_city','business_state','business_location','business_phone_number','inspection_id','violation_id','Neighborhoods (old)','Police Districts','Supervisor Districts','Fire Prevention Districts','Analysis Neighborhoods','business_latitude','business_longitude','inspection_score','inspection_type','Zip Codes'], axis=1)
violations_df.dropna(subset=['business_postal_code','violation_description'],inplace=True)
violations_df.reset_index(inplace=True, drop=True)

We will merge the neighborhood dataset and the violations one so we keep only data pertinent to the neighborhoods under analysis

In [10]:
sf_df = sf_df.merge(violations_df, left_on='Zip Code', right_on='business_postal_code')
sf_df.drop (['business_postal_code'], axis=1, inplace=True)
sf_df

Zip Code                             Neighborhood  Latitude  Longitude  \
0        94102  Hayes Valley/Tenderloin/North of Market     37.78    -122.42   
1        94102  Hayes Valley/Tenderloin/North of Market     37.78    -122.42   
2        94102  Hayes Valley/Tenderloin/North of Market     37.78    -122.42   
3        94102  Hayes Valley/Tenderloin/North of Market     37.78    -122.42   
4        94102  Hayes Valley/Tenderloin/North of Market     37.78    -122.42   
...        ...                                      ...       ...        ...   
11265    94134              Visitacion Valley/Sunnydale     37.72    -122.41   
11266    94134              Visitacion Valley/Sunnydale     37.72    -122.41   
11267    94134              Visitacion Valley/Sunnydale     37.72    -122.41   
11268    94134              Visitacion Valley/Sunnydale     37.72    -122.41   
11269    94134              Visitacion Valley/Sunnydale     37.72    -122.41   

       business_id          business_name       business_address  \
0            82225  Harry Harringtons Pub          460 Larkin St   
1            69705          Daniel's Cafe           154 Ellis St   
2            93072          Subway #68745  170 O'Farrell St #428   
3            69705          Daniel's Cafe           154 Ellis St   
4            72171           John's Grill            63 Ellis St   
...            ...                    ...                    ...   
11265          379            GIANT SCOOP     2708 SAN BRUNO Ave   
11266         2622   WING HING RESTAURANT     2550 SAN BRUNO Ave   
11267         2785       G & L RESTAURANT         198 LELAND Ave   
11268          194    BAYSIDE COFFEE SHOP     2011 BAYSHORE Blvd   
11269        80233        Buffalo Kitchen         107 Leland Ave   

      inspection_date                              violation_description  \
0          2019-03-19                     Improper or defective plumbing   
1          2019-02-12             Moderate risk food holding temperature   
2          2019-01-03                     Improper or defective plumbing   
3          2019-02-12  Wiping cloths not clean or properly stored or ...   
4          2019-01-22                           Improper cooling methods   
...               ...                                                ...   
11265      2019-04-08   Unapproved or unmaintained equipment or utensils   
11266      2019-01-24       Unclean or degraded floors walls or ceilings   
11267      2019-05-02   Unapproved or unmaintained equipment or utensils   
11268      2019-03-19  Inadequate and inaccessible handwashing facili...   
11269      2019-04-17             Foods not protected from contamination   

       risk_category  
0           Low Risk  
1      Moderate Risk  
2           Low Risk  
3           Low Risk  
4          High Risk  
...              ...  
11265       Low Risk  
11266       Low Risk  
11267       Low Risk  
11268  Moderate Risk  
11269  Moderate Risk  

[11270 rows x 10 columns]

Let us Explore different views of the data. Let's begin by understanding the breakdown of violation severity:

In [11]:
explore_df = sf_df.groupby('violation_description')['risk_category'].value_counts().unstack().fillna(0)
sum_column = explore_df["High Risk"] + explore_df["Moderate Risk"] + explore_df["Low Risk"]
explore_df["Total"] = sum_column
explore_df.sort_values(['Total'], ascending=False, inplace=True)
explore_df.head(10)

risk_category                                       High Risk  Low Risk  \
violation_description                                                     
Inadequately cleaned or sanitized food contact ...        0.0       0.0   
Unapproved or unmaintained equipment or utensils          0.0     896.0   
Unclean or degraded floors walls or ceilings              0.0     847.0   
Moderate risk food holding temperature                    0.0       0.0   
Inadequate and inaccessible handwashing facilities        0.0       0.0   
Foods not protected from contamination                    0.0       0.0   
Wiping cloths not clean or properly stored or i...        0.0     526.0   
High risk food holding temperature                      447.0       0.0   
Moderate risk vermin infestation                          0.0       0.0   
Improper food storage                                     0.0     408.0   

risk_category                                       Moderate Risk  Total  
violation_description                                                     
Inadequately cleaned or sanitized food contact ...          961.0  961.0  
Unapproved or unmaintained equipment or utensils              0.0  896.0  
Unclean or degraded floors walls or ceilings                  0.0  847.0  
Moderate risk food holding temperature                      766.0  766.0  
Inadequate and inaccessible handwashing facilities          760.0  760.0  
Foods not protected from contamination                      560.0  560.0  
Wiping cloths not clean or properly stored or i...            0.0  526.0  
High risk food holding temperature                            0.0  447.0  
Moderate risk vermin infestation                            443.0  443.0  
Improper food storage                                         0.0  408.0

In [12]:
explore_df.shape

(62, 4)

In [13]:
explore_df.sum()

risk_category
High Risk         1698.0
Low Risk          5102.0
Moderate Risk     4470.0
Total            11270.0
dtype: float64

In [14]:
explore_df.to_csv('explore.csv',index=True)

There are 62 violation types in our dataframe. A total of 11,270 records. 45% of violations in 2019 are considered low risk, 40% moderate and ~15% high risk. 

Most common violations are related to cleaning standards and food temperature holding. These findings will be relevant to determine solution packages we can offer to potential customers. As well as to understand areas of specialization we need to look for in potential hires joining the company.

# Step 4 : Analyze Neighborhoods and Violations

We will now shift our focus to each neighborhood and violatiosn in them. In order to do so we will group by violation description as shown below:

In [15]:
clustering_df= sf_df.groupby('Neighborhood')['violation_description'].value_counts().unstack().fillna(0)
clustering_df.head(10)

violation_description                    Consumer advisory not provided for raw or undercooked foods  \
Neighborhood                                                                                           
Bayview-Hunters Point                                                                  0.0             
Castro/Noe Valley                                                                      1.0             
Chinatown                                                                              1.0             
Haight-Ashbury                                                                         0.0             
Hayes Valley/Tenderloin/North of Market                                                2.0             
Ingelside-Excelsior/Crocker-Amazon                                                     0.0             
Inner Mission/Bernal Heights                                                           0.0             
Inner Richmond                                                                         0.0             
Lake Merced                                                                            0.0             
Marina                                                                                 0.0             

violation_description                    Contaminated or adulterated food  \
Neighborhood                                                                
Bayview-Hunters Point                                                 0.0   
Castro/Noe Valley                                                     0.0   
Chinatown                                                             1.0   
Haight-Ashbury                                                        0.0   
Hayes Valley/Tenderloin/North of Market                               6.0   
Ingelside-Excelsior/Crocker-Amazon                                    1.0   
Inner Mission/Bernal Heights                                         11.0   
Inner Richmond                                                        1.0   
Lake Merced                                                           0.0   
Marina                                                                2.0   

violation_description                    Discharge from employee nose mouth or eye  \
Neighborhood                                                                         
Bayview-Hunters Point                                                          0.0   
Castro/Noe Valley                                                              0.0   
Chinatown                                                                      0.0   
Haight-Ashbury                                                                 0.0   
Hayes Valley/Tenderloin/North of Market                                        0.0   
Ingelside-Excelsior/Crocker-Amazon                                             0.0   
Inner Mission/Bernal Heights                                                   1.0   
Inner Richmond                                                                 0.0   
Lake Merced                                                                    0.0   
Marina                                                                         0.0   

violation_description                    Employee eating or smoking  \
Neighborhood                                                          
Bayview-Hunters Point                                           2.0   
Castro/Noe Valley                                               2.0   
Chinatown                                                       2.0   
Haight-Ashbury                                                  2.0   
Hayes Valley/Tenderloin/North of Market                         3.0   
Ingelside-Excelsior/Crocker-Amazon                              7.0   
Inner Mission/Bernal Heights                                   14.0   
Inner Richmond                                                  6.0   
Lake Merced                                                     0.0   
Marina      

We will now perform one hot encoding on the grouped data frame

In [16]:
#one hot encoding
sf_onehot=pd.get_dummies(sf_df[['violation_description']],prefix="", prefix_sep="")
#add neighborhood column back to the dataframe, move it ot the first column
sf_onehot['Neighborhood'] = sf_df['Neighborhood']
fixed_columns = [sf_onehot.columns[-1]] + list(sf_onehot.columns[:-1])
sf_onehot = sf_onehot[fixed_columns]

sf_onehot.head()

Neighborhood  \
0  Hayes Valley/Tenderloin/North of Market   
1  Hayes Valley/Tenderloin/North of Market   
2  Hayes Valley/Tenderloin/North of Market   
3  Hayes Valley/Tenderloin/North of Market   
4  Hayes Valley/Tenderloin/North of Market   

   Consumer advisory not provided for raw or undercooked foods  \
0                                                  0             
1                                                  0             
2                                                  0             
3                                                  0             
4                                                  0             

   Contaminated or adulterated food  \
0                                 0   
1                                 0   
2                                 0   
3                                 0   
4                                 0   

   Discharge from employee nose mouth or eye  Employee eating or smoking  \
0                                          0                           0   
1                                          0                           0   
2                                          0                           0   
3                                          0                           0   
4                                          0                           0   

   Food in poor condition  \
0                       0   
1                       0   
2                       0   
3                       0   
4                       0   

   Food safety certificate or food handler card not available  \
0                                                  0            
1                                                  0            
2                                                  0            
3                                                  0            
4                                                  0            

   Foods not protected from contamination  High risk food holding temperature  \
0                                       0                                   0   
1                                       0                                   0   
2                                       0                                   0   
3                                       0                                   0   
4                                       0                                   0   

   High risk vermin infestation  ...  \
0                             0  ...   
1                             0  ...   
2                             0  ...   
3                             0  ...   
4                             0  ...   

   Unauthorized or unsafe use of time as a public health control measure  \
0                                                  0                       
1                                                  0                       
2                                                  0                       
3                                                  0                       
4                                                  0                       

   Unclean hands or improper use of gloves  Unclean nonfood contact surfaces  \
0                                        0                                 0   
1                                        0                                 0   
2                                        0                                 0   
3                                        0                                 0   
4                                        0                                 0   

   Unclean or degraded floors walls or ceilings  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   

   Unclean or unsanitary food contact surfaces  \
0                                            0   
1                        

Take the one hot encoded dataframe, group by mean and reset the index:

In [17]:
sf_grouped = sf_onehot.groupby('Neighborhood').mean().reset_index()
sf_grouped.head()

Neighborhood  \
0                    Bayview-Hunters Point   
1                        Castro/Noe Valley   
2                                Chinatown   
3                           Haight-Ashbury   
4  Hayes Valley/Tenderloin/North of Market   

   Consumer advisory not provided for raw or undercooked foods  \
0                                           0.000000             
1                                           0.003460             
2                                           0.001678             
3                                           0.000000             
4                                           0.002372             

   Contaminated or adulterated food  \
0                          0.000000   
1                          0.000000   
2                          0.001678   
3                          0.000000   
4                          0.007117   

   Discharge from employee nose mouth or eye  Employee eating or smoking  \
0                                        0.0                    0.005291   
1                                        0.0                    0.006920   
2                                        0.0                    0.003356   
3                                        0.0                    0.006849   
4                                        0.0                    0.003559   

   Food in poor condition  \
0                0.005291   
1                0.003460   
2                0.000000   
3                0.000000   
4                0.003559   

   Food safety certificate or food handler card not available  \
0                                           0.042328            
1                                           0.038062            
2                                           0.025168            
3                                           0.044521            
4                                           0.037960            

   Foods not protected from contamination  High risk food holding temperature  \
0                                0.055556                            0.042328   
1                                0.072664                            0.044983   
2                                0.040268                            0.048658   
3                                0.041096                            0.034247   
4                                0.032028                            0.049822   

   High risk vermin infestation  ...  \
0                      0.010582  ...   
1                      0.006920  ...   
2                      0.036913  ...   
3                      0.027397  ...   
4                      0.021352  ...   

   Unauthorized or unsafe use of time as a public health control measure  \
0                                           0.000000                       
1                                           0.000000                       
2                                           0.001678                       
3                                           0.000000                       
4                                           0.000000                       

   Unclean hands or improper use of gloves  Unclean nonfood contact surfaces  \
0                                 0.026455                          0.076720   
1                                 0.003460                          0.006920   
2                                 0.013423                          0.035235   
3                                 0.006849                          0.010274   
4                                 0.005931                          0.046263   

   Unclean or degraded floors walls or ceilings  \
0                                      0.074074   
1                                      0.096886   
2                                      0.058725   
3                                      0.099315   
4                                      0.066429   

   Unclean or unsanitary food contact surfaces  \
0                                     0.007937   
1                        

Function to return the most common violations in each neighborhood:

In [18]:
def return_most_common_violations(row, num_top):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top]

In [19]:
num_top = 10
indicators = ['st','nd','rd']

#create columns according to number of top violations
columns=['Neighborhood']
for ind in np.arange(num_top):
    try:
        #append indicators to top 3 venues
        columns.append('{}{} Most Common Violation'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Violation'.format(ind+1))
        
#set up the new dataframe
neighborhoods_violations_sorted = pd.DataFrame(columns=columns)
neighborhoods_violations_sorted['Neighborhood'] = sf_grouped['Neighborhood']

for ind  in np.arange(sf_grouped.shape[0]):
    neighborhoods_violations_sorted.iloc[ind,1:] = return_most_common_violations(sf_grouped.iloc[ind,:],num_top)
    
neighborhoods_violations_sorted.head()

Neighborhood  \
0                    Bayview-Hunters Point   
1                        Castro/Noe Valley   
2                                Chinatown   
3                           Haight-Ashbury   
4  Hayes Valley/Tenderloin/North of Market   

                           1st Most Common Violation  \
0   Unapproved or unmaintained equipment or utensils   
1  Wiping cloths not clean or properly stored or ...   
2  Inadequately cleaned or sanitized food contact...   
3   Unapproved or unmaintained equipment or utensils   
4  Inadequate and inaccessible handwashing facili...   

                           2nd Most Common Violation  \
0  Inadequately cleaned or sanitized food contact...   
1       Unclean or degraded floors walls or ceilings   
2  Inadequate and inaccessible handwashing facili...   
3       Unclean or degraded floors walls or ceilings   
4             Moderate risk food holding temperature   

                           3rd Most Common Violation  \
0                   Unclean nonfood contact surfaces   
1  Inadequately cleaned or sanitized food contact...   
2   Unapproved or unmaintained equipment or utensils   
3             Moderate risk food holding temperature   
4  Inadequately cleaned or sanitized food contact...   

                          4th Most Common Violation  \
0      Unclean or degraded floors walls or ceilings   
1            Moderate risk food holding temperature   
2            Moderate risk food holding temperature   
3                  Moderate risk vermin infestation   
4  Unapproved or unmaintained equipment or utensils   

                           5th Most Common Violation  \
0             Foods not protected from contamination   
1  Inadequate and inaccessible handwashing facili...   
2                              Improper food storage   
3  Inadequately cleaned or sanitized food contact...   
4       Unclean or degraded floors walls or ceilings   

                           6th Most Common Violation  \
0             Moderate risk food holding temperature   
1             Foods not protected from contamination   
2       Unclean or degraded floors walls or ceilings   
3  Inadequate and inaccessible handwashing facili...   
4                 High risk food holding temperature   

                           7th Most Common Violation  \
0  Inadequate and inaccessible handwashing facili...   
1   Unapproved or unmaintained equipment or utensils   
2                 High risk food holding temperature   
3        Unclean or unsanitary food contact surfaces   
4  Wiping cloths not clean or properly stored or ...   

                           8th Most Common Violation  \
0                           Improper cooling methods   
1                   Moderate risk vermin infestation   
2  Wiping cloths not clean or properly stored or ...   
3  Food safety certificate or food handler card n...   
4                   Unclean nonfood contact surfaces   

                           9th Most Common Violation  \
0  Wiping cloths not clean or properly stored or ...   
1                        Low risk vermin infestation   
2             Foods not protected from contamination   
3             Foods not protected from contamination   
4                              Improper food storage   

                         10th Most Common Violation  
0                       Low risk vermin infestation  
1                High risk food holding temperature  
2                       Low risk vermin infestation  
3  Improper storage of equipment utensils or linens  
4                          Improper cooling methods

# Step 5 : Clustering neighborhoods according to violations

We will now apply the k means algorithm to the neighborhoods to be able to cluster them based on the most common violations:

In [20]:
from sklearn.cluster import KMeans
#set number of clusters
kclusters = 5
sf_grouped_clustering = sf_grouped.drop('Neighborhood',1)
#run k-means clustering
kmeans = KMeans (n_clusters = kclusters, random_state=0).fit(sf_grouped_clustering)
#check cluster labels
kmeans.labels_[0:10]

array([0, 1, 2, 4, 2, 2, 2, 4, 1, 2], dtype=int32)

Add the cluster labels to the original dataframe

In [21]:
neighborhoods_violations_sorted.insert(0,'Cluster Labels', kmeans.labels_)
sf_merged = sf_neigh
sf_merged = sf_merged.merge(neighborhoods_violations_sorted, on ='Neighborhood')
sf_merged.head()

Zip Code                             Neighborhood  Latitude  Longitude  \
0    94102  Hayes Valley/Tenderloin/North of Market    37.780   -122.420   
1    94103                          South of Market    37.780   -122.410   
2    94107                             Potrero Hill    37.770   -122.390   
3    94108                                Chinatown    37.791   -122.409   
4    94109             Polk/Russian Hill (Nob Hill)    37.790   -122.420   

   Cluster Labels                          1st Most Common Violation  \
0               2  Inadequate and inaccessible handwashing facili...   
1               2       Unclean or degraded floors walls or ceilings   
2               2             Moderate risk food holding temperature   
3               2  Inadequately cleaned or sanitized food contact...   
4               2   Unapproved or unmaintained equipment or utensils   

                           2nd Most Common Violation  \
0             Moderate risk food holding temperature   
1  Inadequately cleaned or sanitized food contact...   
2  Inadequately cleaned or sanitized food contact...   
3  Inadequate and inaccessible handwashing facili...   
4  Inadequate and inaccessible handwashing facili...   

                           3rd Most Common Violation  \
0  Inadequately cleaned or sanitized food contact...   
1             Moderate risk food holding temperature   
2   Unapproved or unmaintained equipment or utensils   
3   Unapproved or unmaintained equipment or utensils   
4  Inadequately cleaned or sanitized food contact...   

                           4th Most Common Violation  \
0   Unapproved or unmaintained equipment or utensils   
1   Unapproved or unmaintained equipment or utensils   
2  Inadequate and inaccessible handwashing facili...   
3             Moderate risk food holding temperature   
4             Moderate risk food holding temperature   

                           5th Most Common Violation  \
0       Unclean or degraded floors walls or ceilings   
1  Inadequate and inaccessible handwashing facili...   
2       Unclean or degraded floors walls or ceilings   
3                              Improper food storage   
4  Wiping cloths not clean or properly stored or ...   

                      6th Most Common Violation  \
0            High risk food holding temperature   
1            High risk food holding temperature   
2        Foods not protected from contamination   
3  Unclean or degraded floors walls or ceilings   
4  Unclean or degraded floors walls or ceilings   

                           7th Most Common Violation  \
0  Wiping cloths not clean or properly stored or ...   
1             Foods not protected from contamination   
2  Food safety certificate or food handler card n...   
3                 High risk food holding temperature   
4             Foods not protected from contamination   

                           8th Most Common Violation  \
0                   Unclean nonfood contact surfaces   
1                   Moderate risk vermin infestation   
2  Wiping cloths not clean or properly stored or ...   
3  Wiping cloths not clean or properly stored or ...   
4                              Improper food storage   

                           9th Most Common Violation  \
0                              Improper food storage   
1  Wiping cloths not clean or properly stored or ...   
2                 High risk food holding temperature   
3             Foods not protected from contamination   
4                   Unclean nonfood contact surfaces   

                          10th Most Common Violation  
0                           Improper cooling methods  
1  Food safety certificate or food handler card n...  
2  Inadequate food safety knowledge or lack of ce...  
3                        Low risk vermin infestation  
4                 High risk food holding temperature

Let's explore the defining characteristics of each cluster:

Cluster 1:

In [22]:
sf_merged.loc[sf_merged['Cluster Labels'] == 0, sf_merged.columns[[1] + list(range(5, sf_merged.shape[1]))]]

Neighborhood  \
15        Bayview-Hunters Point   
20  Visitacion Valley/Sunnydale   

                           1st Most Common Violation  \
15  Unapproved or unmaintained equipment or utensils   
20  Unapproved or unmaintained equipment or utensils   

                            2nd Most Common Violation  \
15  Inadequately cleaned or sanitized food contact...   
20                 High risk food holding temperature   

                            3rd Most Common Violation  \
15                   Unclean nonfood contact surfaces   
20  Inadequately cleaned or sanitized food contact...   

                            4th Most Common Violation  \
15       Unclean or degraded floors walls or ceilings   
20  Inadequate and inaccessible handwashing facili...   

                 5th Most Common Violation  \
15  Foods not protected from contamination   
20        Unclean nonfood contact surfaces   

                 6th Most Common Violation  \
15  Moderate risk food holding temperature   
20  Moderate risk food holding temperature   

                            7th Most Common Violation  \
15  Inadequate and inaccessible handwashing facili...   
20       Unclean or degraded floors walls or ceilings   

                  8th Most Common Violation  \
15                 Improper cooling methods   
20  Unclean hands or improper use of gloves   

                            9th Most Common Violation  \
15  Wiping cloths not clean or properly stored or ...   
20                           Improper thawing methods   

                           10th Most Common Violation  
15                        Low risk vermin infestation  
20  Wiping cloths not clean or properly stored or ...

Cluster 2:

In [23]:
sf_merged.loc[sf_merged['Cluster Labels'] == 1, sf_merged.columns[[1] + list(range(5, sf_merged.shape[1]))]]

Neighborhood  \
7                       Castro/Noe Valley   
16  St. Francis Wood/Miraloma/West Portal   
17                   Twin Peaks-Glen Park   
18                            Lake Merced   

                            1st Most Common Violation  \
7   Wiping cloths not clean or properly stored or ...   
16             Foods not protected from contamination   
17             Foods not protected from contamination   
18  Wiping cloths not clean or properly stored or ...   

                            2nd Most Common Violation  \
7        Unclean or degraded floors walls or ceilings   
16       Unclean or degraded floors walls or ceilings   
17  Wiping cloths not clean or properly stored or ...   
18  Inadequate and inaccessible handwashing facili...   

                            3rd Most Common Violation  \
7   Inadequately cleaned or sanitized food contact...   
16             Moderate risk food holding temperature   
17       Unclean or degraded floors walls or ceilings   
18             Foods not protected from contamination   

                            4th Most Common Violation  \
7              Moderate risk food holding temperature   
16  Inadequately cleaned or sanitized food contact...   
17             Moderate risk food holding temperature   
18       Unclean or degraded floors walls or ceilings   

                            5th Most Common Violation  \
7   Inadequate and inaccessible handwashing facili...   
16   Unapproved or unmaintained equipment or utensils   
17   Unapproved or unmaintained equipment or utensils   
18     Permit license or inspection report not posted   

                            6th Most Common Violation  \
7              Foods not protected from contamination   
16  Wiping cloths not clean or properly stored or ...   
17  Food safety certificate or food handler card n...   
18  Inadequately cleaned or sanitized food contact...   

                           7th Most Common Violation  \
7   Unapproved or unmaintained equipment or utensils   
16                          Improper thawing methods   
17                       Low risk vermin infestation   
18            Moderate risk food holding temperature   

                      8th Most Common Violation  \
7              Moderate risk vermin infestation   
16             Unclean nonfood contact surfaces   
17  Unclean or unsanitary food contact surfaces   
18                     Improper thawing methods   

                            9th Most Common Violation  \
7                         Low risk vermin infestation   
16  Inadequate and inaccessible handwashing facili...   
17  Inadequately cleaned or sanitized food contact...   
18                     Improper or defective plumbing   

                     10th Most Common Violation  
7            High risk food holding temperature  
16                  Low risk vermin infestation  
17                     Improper thawing methods  
18  Unclean or unsanitary food contact surfaces

Cluster 3:

In [24]:
sf_merged.loc[sf_merged['Cluster Labels'] == 2, sf_merged.columns[[1] + list(range(5, sf_merged.shape[1]))]]

Neighborhood  \
0   Hayes Valley/Tenderloin/North of Market   
1                           South of Market   
2                              Potrero Hill   
3                                 Chinatown   
4              Polk/Russian Hill (Nob Hill)   
5              Inner Mission/Bernal Heights   
6        Ingelside-Excelsior/Crocker-Amazon   
14                                   Marina   
19                    North Beach/Chinatown   

                            1st Most Common Violation  \
0   Inadequate and inaccessible handwashing facili...   
1        Unclean or degraded floors walls or ceilings   
2              Moderate risk food holding temperature   
3   Inadequately cleaned or sanitized food contact...   
4    Unapproved or unmaintained equipment or utensils   
5   Inadequately cleaned or sanitized food contact...   
6   Inadequately cleaned or sanitized food contact...   
14  Food safety certificate or food handler card n...   
19  Inadequately cleaned or sanitized food contact...   

                            2nd Most Common Violation  \
0              Moderate risk food holding temperature   
1   Inadequately cleaned or sanitized food contact...   
2   Inadequately cleaned or sanitized food contact...   
3   Inadequate and inaccessible handwashing facili...   
4   Inadequate and inaccessible handwashing facili...   
5        Unclean or degraded floors walls or ceilings   
6    Unapproved or unmaintained equipment or utensils   
14   Unapproved or unmaintained equipment or utensils   
19  Inadequate and inaccessible handwashing facili...   

                            3rd Most Common Violation  \
0   Inadequately cleaned or sanitized food contact...   
1              Moderate risk food holding temperature   
2    Unapproved or unmaintained equipment or utensils   
3    Unapproved or unmaintained equipment or utensils   
4   Inadequately cleaned or sanitized food contact...   
5              Moderate risk food holding temperature   
6        Unclean or degraded floors walls or ceilings   
14       Unclean or degraded floors walls or ceilings   
19   Unapproved or unmaintained equipment or utensils   

                            4th Most Common Violation  \
0    Unapproved or unmaintained equipment or utensils   
1    Unapproved or unmaintained equipment or utensils   
2   Inadequate and inaccessible handwashing facili...   
3              Moderate risk food holding temperature   
4              Moderate risk food holding temperature   
5    Unapproved or unmaintained equipment or utensils   
6   Inadequate and inaccessible handwashing facili...   
14  Inadequately cleaned or sanitized food contact...   
19             Moderate risk food holding temperature   

                            5th Most Common Violation  \
0        Unclean or degraded floors walls or ceilings   
1   Inadequate and inaccessible handwashing facili...   
2        Unclean or degraded floors walls or ceilings   
3                               Improper food storage   
4   Wiping cloths not clean or properly stored or ...   
5                    Unclean nonfood contact surfaces   
6                  High risk food holding temperature   
14  Wiping cloths not clean or properly stored or ...   
19       Unclean or degraded floors walls or ceilings   

                            6th Most Common Violation  \
0                  High risk food holding temperature   
1                  High risk food holding temperature   
2              Foods not protected from contamination   
3        Unclean or degraded floors walls or ceilings   
4        Unclean or degraded floors walls or ceilings   
5                    Moderate risk vermin infestation   
6              Moderate risk food holding temperature   
14  Inadequate and inaccessible handwashing facili...   
19   Improper storage of equipment utensils or linens   

                            7th Most Common Violation  \
0   Wiping cloths not clean or properly stored or ...   
1              Foods

Cluster 4:

In [25]:
sf_merged.loc[sf_merged['Cluster Labels'] == 3, sf_merged.columns[[1] + list(range(5, sf_merged.shape[1]))]]

Neighborhood                     1st Most Common Violation  \
9   Parkside/Forest Hill  Unclean or degraded floors walls or ceilings   
13                Sunset  Unclean or degraded floors walls or ceilings   

                 2nd Most Common Violation  \
9   Foods not protected from contamination   
13  Foods not protected from contamination   

                            3rd Most Common Violation  \
9   Inadequate and inaccessible handwashing facili...   
13  Inadequate and inaccessible handwashing facili...   

                            4th Most Common Violation  \
9   Inadequately cleaned or sanitized food contact...   
13             Moderate risk food holding temperature   

                            5th Most Common Violation  \
9                               Improper food storage   
13  Wiping cloths not clean or properly stored or ...   

                            6th Most Common Violation  \
9                    Moderate risk vermin infestation   
13  Inadequately cleaned or sanitized food contact...   

       7th Most Common Violation           8th Most Common Violation  \
9   High risk vermin infestation    Unclean nonfood contact surfaces   
13         Improper food storage  High risk food holding temperature   

                           9th Most Common Violation  \
9   Unapproved or unmaintained equipment or utensils   
13                      High risk vermin infestation   

                10th Most Common Violation  
9   Moderate risk food holding temperature  
13        Moderate risk vermin infestation

Cluster 5:

In [26]:
sf_merged.loc[sf_merged['Cluster Labels'] == 4, sf_merged.columns[[1] + list(range(5, sf_merged.shape[1]))]]

Neighborhood  \
8   Western Addition/Japantown   
10              Haight-Ashbury   
11              Inner Richmond   
12              Outer Richmond   

                            1st Most Common Violation  \
8    Unapproved or unmaintained equipment or utensils   
10   Unapproved or unmaintained equipment or utensils   
11   Unapproved or unmaintained equipment or utensils   
12  Inadequately cleaned or sanitized food contact...   

                            2nd Most Common Violation  \
8   Inadequately cleaned or sanitized food contact...   
10       Unclean or degraded floors walls or ceilings   
11  Inadequately cleaned or sanitized food contact...   
12   Unapproved or unmaintained equipment or utensils   

                            3rd Most Common Violation  \
8                    Moderate risk vermin infestation   
10             Moderate risk food holding temperature   
11        Unclean or unsanitary food contact surfaces   
12  Inadequate and inaccessible handwashing facili...   

                            4th Most Common Violation  \
8        Unclean or degraded floors walls or ceilings   
10                   Moderate risk vermin infestation   
11  Inadequate and inaccessible handwashing facili...   
12             Moderate risk food holding temperature   

                            5th Most Common Violation  \
8   Inadequate and inaccessible handwashing facili...   
10  Inadequately cleaned or sanitized food contact...   
11             Moderate risk food holding temperature   
12                           Improper thawing methods   

                            6th Most Common Violation  \
8   Wiping cloths not clean or properly stored or ...   
10  Inadequate and inaccessible handwashing facili...   
11                   Moderate risk vermin infestation   
12  Wiping cloths not clean or properly stored or ...   

                       7th Most Common Violation  \
8         Moderate risk food holding temperature   
10   Unclean or unsanitary food contact surfaces   
11  Unclean or degraded floors walls or ceilings   
12                         Improper food storage   

                            8th Most Common Violation  \
8                         Low risk vermin infestation   
10  Food safety certificate or food handler card n...   
11                              Improper food storage   
12        Unclean or unsanitary food contact surfaces   

                 9th Most Common Violation  \
8   Foods not protected from contamination   
10  Foods not protected from contamination   
11  Foods not protected from contamination   
12        Moderate risk vermin infestation   

                          10th Most Common Violation  
8                              Improper food storage  
10  Improper storage of equipment utensils or linens  
11                          Improper thawing methods  
12                       Low risk vermin infestation

Now, we will plot the clusters in the map

In [27]:
# create map
map_clusters = folium.Map(location = [latitude, longitude], zoom_start = 12, height=500, width=600)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i * x) ** 2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(sf_merged['Latitude'], sf_merged['Longitude'], sf_merged['Neighborhood'], sf_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius = 10,
        popup = label,
        tooltip=label,
        color = rainbow[cluster - 1],
        fill = True,
        fill_color = rainbow[cluster - 1],
        fill_opacity = 0.7).add_to(map_clusters)
       
map_clusters


# Step 6 : Exploring the market with Foursquare

In order to explore the marketplace and find the most common food places in each neighbohood we will use the Foursquare API:

In [28]:
from kaggle_secrets import UserSecretsClient
user_secrets = UserSecretsClient()
secret_value_0 = user_secrets.get_secret("CLIENT_ID")
secret_value_1 = user_secrets.get_secret("CLIENT_SECRET")

CLIENT_ID = user_secrets.get_secret("CLIENT_ID") # your Foursquare ID
CLIENT_SECRET = user_secrets.get_secret("CLIENT_SECRET") # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100

Function to pull all venues within a 500m radius. By passing the 'Food' Category ID, we're filtering only food places at the API level:

In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&categoryId=4d4b7105d754a06374d81259&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    return(nearby_venues)

In [30]:
sf_venues = getNearbyVenues(names = sf_neigh['Neighborhood'],
                                   latitudes = sf_neigh['Latitude'],
                                   longitudes = sf_neigh['Longitude']
                                  )
                                  
sf_venues.shape

(690, 7)

Let's look at how many venues we have for each neighborhood:

In [31]:
sf_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                     
Bayview-Hunters Point                                        3   
Castro/Noe Valley                                           31   
Chinatown                                                   41   
Haight-Ashbury                                              13   
Hayes Valley/Tenderloin/North of Market                     57   
Ingelside-Excelsior/Crocker-Amazon                          42   
Inner Mission/Bernal Heights                                38   
Inner Richmond                                              45   
Lake Merced                                                 16   
Marina                                                      51   
North Beach/Chinatown                                       51   
Outer Richmond                                              25   
Parkside/Forest Hill                                        37   
Polk/Russian Hill (Nob Hill)                                58   
Potrero Hill                                                55   
South of Market                                             47   
Sunset                                                      52   
Twin Peaks-Glen Park                                        10   
Visitacion Valley/Sunnydale                                  1   
Western Addition/Japantown                                  17   

                                         Neighborhood Longitude  Venue  \
Neighborhood                                                             
Bayview-Hunters Point                                         3      3   
Castro/Noe Valley                                            31     31   
Chinatown                                                    41     41   
Haight-Ashbury                                               13     13   
Hayes Valley/Tenderloin/North of Market                      57     57   
Ingelside-Excelsior/Crocker-Amazon                           42     42   
Inner Mission/Bernal Heights                                 38     38   
Inner Richmond                                               45     45   
Lake Merced                                                  16     16   
Marina                                                       51     51   
North Beach/Chinatown                                        51     51   
Outer Richmond                                               25     25   
Parkside/Forest Hill                                         37     37   
Polk/Russian Hill (Nob Hill)                                 58     58   
Potrero Hill                                                 55     55   
South of Market                                              47     47   
Sunset                                                       52     52   
Twin Peaks-Glen Park                                         10     10   
Visitacion Valley/Sunnydale                                   1      1   
Western Addition/Japantown                                   17     17   

                                         Venue Latitude  Venue Longitude  \
Neighborhood                                                               
Bayview-Hunters Point                                 3                3   
Castro/Noe Valley                                    31               31   
Chinatown                                            41               41   
Haight-Ashbury                                       13               13   
Hayes Valley/Tenderloin/North of Market              57               57   
Ingelside-Excelsior/Crocker-Amazon                   42               42   
Inner Mission/Bernal Heights                         38               38   
Inner Richmond                                       45               45   
Lake Merced                                          16               16   
Marina                                               51               51   
North Beach/Chinatown                          

In [32]:
print('There are {} uniques categories.'.format(len(sf_venues['Venue Category'].unique())))

There are 76 uniques categories.


We need to turn the venue category into dummy variables

In [33]:
# one hot encoding
sf_onehot = pd.get_dummies(sf_venues[['Venue Category']], prefix = "", prefix_sep = "")

# add neighborhood column back to dataframe
sf_onehot['Neighborhood'] = sf_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [sf_onehot.columns[-1]] + list(sf_onehot.columns[:-1])
sf_onehot = sf_onehot[fixed_columns]

sf_onehot.head()

Neighborhood  Afghan Restaurant  \
0  Hayes Valley/Tenderloin/North of Market                  0   
1  Hayes Valley/Tenderloin/North of Market                  0   
2  Hayes Valley/Tenderloin/North of Market                  0   
3  Hayes Valley/Tenderloin/North of Market                  0   
4  Hayes Valley/Tenderloin/North of Market                  0   

   American Restaurant  Asian Restaurant  BBQ Joint  Bakery  \
0                    0                 0          0       0   
1                    0                 0          0       0   
2                    0                 0          0       0   
3                    0                 0          0       0   
4                    0                 0          0       0   

   Brazilian Restaurant  Breakfast Spot  Burger Joint  Burmese Restaurant  \
0                     0               0             0                   0   
1                     0               0             0                   0   
2                     0               0             0                   0   
3                     0               0             0                   0   
4                     0               0             0                   0   

   ...  Southern / Soul Food Restaurant  Spanish Restaurant  Steakhouse  \
0  ...                                0                   0           0   
1  ...                                0                   0           0   
2  ...                                0                   0           0   
3  ...                                1                   0           0   
4  ...                                0                   0           0   

   Sushi Restaurant  Szechuan Restaurant  Taco Place  Thai Restaurant  \
0                 0                    0           0                0   
1                 0                    0           0                0   
2                 0                    0           0                0   
3                 0                    0           0                0   
4                 1                    0           0                0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Vietnamese Restaurant  
0                   0                              0                      0  
1                   0                              0                      0  
2                   0                              0                      0  
3                   0                              0                      0  
4                   0                              0                      0  

[5 rows x 77 columns]

We will now group neighborhoods taking the mean of the frequency of each venue category

In [34]:
sf_grouped = sf_onehot.groupby('Neighborhood').mean().reset_index()
sf_grouped.head()

Neighborhood  Afghan Restaurant  \
0                    Bayview-Hunters Point                0.0   
1                        Castro/Noe Valley                0.0   
2                                Chinatown                0.0   
3                           Haight-Ashbury                0.0   
4  Hayes Valley/Tenderloin/North of Market                0.0   

   American Restaurant  Asian Restaurant  BBQ Joint    Bakery  \
0             0.000000           0.00000        0.0  0.666667   
1             0.032258           0.00000        0.0  0.000000   
2             0.048780           0.02439        0.0  0.024390   
3             0.000000           0.00000        0.0  0.153846   
4             0.052632           0.00000        0.0  0.035088   

   Brazilian Restaurant  Breakfast Spot  Burger Joint  Burmese Restaurant  \
0                   0.0         0.00000      0.000000                 0.0   
1                   0.0         0.00000      0.064516                 0.0   
2                   0.0         0.02439      0.024390                 0.0   
3                   0.0         0.00000      0.000000                 0.0   
4                   0.0         0.00000      0.035088                 0.0   

   ...  Southern / Soul Food Restaurant  Spanish Restaurant  Steakhouse  \
0  ...                         0.000000                 0.0     0.00000   
1  ...                         0.000000                 0.0     0.00000   
2  ...                         0.024390                 0.0     0.04878   
3  ...                         0.000000                 0.0     0.00000   
4  ...                         0.035088                 0.0     0.00000   

   Sushi Restaurant  Szechuan Restaurant  Taco Place  Thai Restaurant  \
0          0.000000             0.000000    0.000000         0.000000   
1          0.032258             0.032258    0.032258         0.096774   
2          0.073171             0.000000    0.000000         0.024390   
3          0.000000             0.000000    0.000000         0.076923   
4          0.052632             0.000000    0.017544         0.070175   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Vietnamese Restaurant  
0                 0.0                       0.000000               0.000000  
1                 0.0                       0.000000               0.000000  
2                 0.0                       0.024390               0.048780  
3                 0.0                       0.000000               0.000000  
4                 0.0                       0.052632               0.087719  

[5 rows x 77 columns]

Function to return most common venues:

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending = False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

We will explore top 4 food venue types for each neighborhood:

In [36]:
num_top_venues = 4

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        # append 'st', 'nd', 'rd' to the top 3 venues
        columns.append('{}{} Most Common Venue'.format(ind + 1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind + 1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns = columns)
neighborhoods_venues_sorted['Neighborhood'] = sf_grouped['Neighborhood']

for ind in np.arange(sf_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(sf_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                     Bayview-Hunters Point                Bakery   
1                         Castro/Noe Valley         Deli / Bodega   
2                                 Chinatown                  Café   
3                            Haight-Ashbury                Bakery   
4   Hayes Valley/Tenderloin/North of Market        Sandwich Place   
5        Ingelside-Excelsior/Crocker-Amazon           Pizza Place   
6              Inner Mission/Bernal Heights    Mexican Restaurant   
7                            Inner Richmond      Sushi Restaurant   
8                               Lake Merced                  Café   
9                                    Marina    Italian Restaurant   
10                    North Beach/Chinatown    Italian Restaurant   
11                           Outer Richmond                  Café   
12                     Parkside/Forest Hill    Chinese Restaurant   
13             Polk/Russian Hill (Nob Hill)    Mexican Restaurant   
14                             Potrero Hill            Food Truck   
15                          South of Market                  Café   
16                                   Sunset    Chinese Restaurant   
17                     Twin Peaks-Glen Park                  Food   
18              Visitacion Valley/Sunnydale   American Restaurant   
19               Western Addition/Japantown    Mexican Restaurant   

       2nd Most Common Venue  3rd Most Common Venue  4th Most Common Venue  
0                       Food  Vietnamese Restaurant    Fried Chicken Joint  
1   Mediterranean Restaurant        Thai Restaurant            Pizza Place  
2         Chinese Restaurant       Sushi Restaurant     Italian Restaurant  
3                       Café      Indian Restaurant          Burrito Place  
4      Vietnamese Restaurant        Thai Restaurant          Deli / Bodega  
5         Chinese Restaurant     Mexican Restaurant  Vietnamese Restaurant  
6              Deli / Bodega     Italian Restaurant                   Café  
7      Vietnamese Restaurant            Pizza Place    Japanese Restaurant  
8                 Food Truck         Sandwich Place            Pizza Place  
9          French Restaurant    American Restaurant        Thai Restaurant  
10         French Restaurant    American Restaurant        Thai Restaurant  
11          Sushi Restaurant     Chinese Restaurant            Pizza Place  
12                      Café            Pizza Place         Sandwich Place  
13          Sushi Restaurant                  Diner     Italian Restaurant  
14                      Café            Pizza Place         Sandwich Place  
15            Sandwich Place             Food Truck             Restaurant  
16                    Bakery  Vietnamese Restaurant    Szechuan Restaurant  
17        Mexican Restaurant    American Restaurant        Thai Restaurant  
18     Vietnamese Restaurant    Fried Chicken Joint   Ethiopian Restaurant  
19        Chinese Restaurant       Sushi Restaurant            Snack Place

In [37]:
neighborhoods_venues_sorted.to_csv('venues.csv',index=False)

# Conclusions
* Identifying most common violations in the city has been possible with the dataset available. Overall business strategy should be planned around the most common ocurrences. This data should be used to drive the marketing strategy along with areas of focus when hiring specialists to be deployed to the different neighborhoods
* Neighborhoods have been clustered based on most common violations and there seems to be a geographical alignment between them (worth digging deeper into possible causes). In any case these geographical alignment can help us tailor the marketing strategy in these neighborhoods in an effective manner
* Varied cuisine is found in San Francisco. Most common venues are Cafés, Italian, Chinese and Mexican Restaurants. Tailoring our offering to these common venues can help us increase market share 


# Additional Considerations
* COVID19 has impacted all industries and restaurants are not the exception, in fact, they have been particularly impacted. Special consideration to the new business landscape needs to be taken, understanding possible regulatory changes as well as financial struggles from restaurant owners which can decrease the appetite for specialized services
* Despite the fact filtering has been made to select ‘Food’ places in the Foursquare API. This API is not necessarily complete as not all restaurants will be registered and additionally other places of interest (e.g. School Cafeterias) will not be returned. Therefore, further market research can be done around other venue types. Out of scope for this exercise